woocommerce/woocommerce-ios
# [Bundle products in order form] Add variation attribute picker UI

Part of #10428 
⚠️ Please review https://github.com/woocommerce/woocommerce-ios/pull/11038 first


## Why


If a bundled item is a variable product with variations that include any attributes with an unspecified "Any" option, then the merchant needs to also manually select an option since all attributes need to have an option in a variable bundle configuration. This PR adds a picker UI in the variable bundle item view, which is shown for these "Any" wildcard attributes.

## How

### Picker UI

The new picker UI is `ConfigurableVariableBundleAttributePicker`, which contains the attribute name and a picker. The view model takes in a `ProductAttribute` from a variable product, and an optional option value which can be set from the product bundle settings in core:

<img width="400" alt="Screenshot 2023-11-01 at 3 08 16 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/b3947488-cc77-4140-9a7f-b6d3c5db7424">

### View model changes for the "Any" wildcard attributes

After a variation is selected from the variation selector, this variation is observed and its attributes are checked against the variable product's full list of attributes to see if there are any missing attributes.

To support updating an existing variable bundled order item, the `selectedVariation: ProductVariation` was changed to a new struct `Variation` with a minimum amount of properties so a remote sync isn't needed.

### Other UI changes:
- In the variable bundle item configuration row, the variation selection CTA now reads `Update variation` when a variation has been selected, otherwise it reads `Select variation`
- A divider is added to the bottom of each bundle item configuration row

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product with non-empty items and **at least one variable item with a variation that has "Any" attribute**.

- Go to the Orders tab
- Tap `+` to create an order
- Tap `Add Products` and select a bundle product in the prerequisite --> the bundle product should be added to the order
- Tap `Create` to create the order first (🗒️ : there's a bug where updating an order with bundle configuration during order creation doesn't set the attributes. I'd like to debug after this PR & https://github.com/woocommerce/woocommerce-ios/pull/11034 are both merged)
- Tap `Edit` to edit the order
- Tap on the expand  🔽 CTA on the bundle product order item
- Tap `Configure` --> after a bit (products being loaded remotely), the configurable bundle items should be shown
- Tap `Select variation` on the variable item and select a variation that has an "Any" attribute --> a picker should be shown for the "Any" attributes. If any default options are set in core, the default options should be shown by default
- Pick a value for all the wildcard attributes
- Tap `Done` --> the attributes should be saved remotely and shown correctly in order details

## Screenshots



https://github.com/woocommerce/woocommerce-ios/assets/1945542/b8e04af8-1ffb-4a80-a3cc-ef03aa132759


Fixed attributes | Attribute picker UI
-- | --
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/98d2176c-aa0b-4dbf-9c11-3c2cc2bd52e5" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/27902794-5fe6-4185-aead-c9c92f1c896c" width="300" />

---
